File: src/main/scala/com/test/Spark2Hbase.scala
package com.test


import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.hbase.client.{Connection, ConnectionFactory}
import org.apache.hadoop.hbase.spark.{ByteArrayWrapper, FamiliesQualifiersValues, FamilyHFileWriteOptions, HBaseContext}
import org.apache.hadoop.hbase.util.Bytes
import org.apache.hadoop.hbase.{HBaseConfiguration, HConstants, TableName}
import org.apache.spark.SparkContext
import org.apache.spark.sql.{DataFrame, SparkSession}

import java.util.HashMap
import java.util.UUID

object Spark2Hbase {
  def main(args: Array[String]): Unit = {
    //hbase org.apache.hadoop.hbase.mapreduce.LoadIncrementalHFiles /tmp/hbase/HFlileOut/ staff
    val sourceTable: String = "staff"
    val spark: SparkSession = SparkSession
      .builder()
      .enableHiveSupport()
      .getOrCreate()
    val sc: SparkContext = spark.sparkContext

    val conf: Configuration = HBaseConfiguration.create()
    conf.set("hbase.zookeeper.quorum", "192.168.7.169:2181")
    //  conf.setInt("zookeeper.recovery.retry", 0)
    //  conf.setInt("hbase.client.retries.number", 0)
    val hbaseContext = new HBaseContext(sc, conf)

    val cf: String = "f1"
    val conn: Connection = ConnectionFactory.createConnection(conf)
    //    val tableName: TableName = createTable(conn, sourceTable, cf)
    //    //先删除可能存在的目录
    //    delete_hdfspath(savePath.toUri.getPath)
    //一定要导入这个包才能使用hbaseBulkLoadThinRows
    import org.apache.hadoop.hbase.spark.HBaseRDDFunctions.GenericHBaseRDDFunctions
    //获取数据
    val sourceDataFrame: DataFrame = spark.sql("select * from ods.customer")
    val columnsName: Array[String] = sourceDataFrame.columns //获取所有列名
    sourceDataFrame
      .rdd
      .map(row => {
        val familyQualifiersValues: FamiliesQualifiersValues = new FamiliesQualifiersValues
        val rowkey: String = UUID.randomUUID.toString
        //对每一列进行处理
        for (i <- 0 until columnsName.length - 1) {
          try {
            familyQualifiersValues += (Bytes.toBytes(cf), Bytes.toBytes(columnsName(i)), Bytes.toBytes(row.getAs(columnsName(i)).toString))
          } catch {
            case e: ClassCastException =>
              familyQualifiersValues += (Bytes.toBytes(cf), Bytes.toBytes(columnsName(i)), Bytes.toBytes(row.getAs(columnsName(i)).toString))
            case e: Exception =>
              e.printStackTrace()
          }
        }
        (new ByteArrayWrapper(Bytes.toBytes(rowkey)), familyQualifiersValues)
      }).hbaseBulkLoadThinRows(hbaseContext, TableName.valueOf(sourceTable), t => t, "/tmp/hbase/HFlileOut", new HashMap[Array[Byte], FamilyHFileWriteOptions], compactionExclude = false, HConstants.DEFAULT_MAX_FILE_SIZE)
    spark.stop()
  }
}

File: src/main/java/com/test/demo/Spark2Hbase.java
package com.test.demo;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.hbase.HBaseConfiguration;
import org.apache.hadoop.hbase.HConstants;
import org.apache.hadoop.hbase.TableName;
import org.apache.hadoop.hbase.spark.*;
import org.apache.hadoop.hbase.util.Bytes;
import org.apache.hadoop.hbase.util.Pair;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

import java.util.HashMap;
import java.util.UUID;


public class Spark2Hbase {
    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("JavaHBaseBulkLoadExample  " + "{outputPath}");
            return;
        }

        String tableName = args[3];
        SparkSession spark = SparkSession
                .builder()
                .appName("Spark2Hbase")
                .enableHiveSupport()
                .getOrCreate();
        Dataset<Row> ds = spark.sql(args[2]);
        String[] columnsNames = ds.columns();
        JavaSparkContext jsc = JavaSparkContext.fromSparkContext(spark.sparkContext());
        Configuration conf = HBaseConfiguration.create();
        conf.set("hbase.zookeeper.quorum", args[1]);
        JavaHBaseContext hbaseContext = new JavaHBaseContext(jsc, conf);

        JavaRDD<Pair<ByteArrayWrapper, FamiliesQualifiersValues>> myrdd = ds.toJavaRDD().map(new Function<Row, Pair<ByteArrayWrapper, FamiliesQualifiersValues>>() {
            @Override
            public Pair<ByteArrayWrapper, FamiliesQualifiersValues> call(Row row) throws Exception {
                FamiliesQualifiersValues familiesQualifiersValues = new FamiliesQualifiersValues();
                for (String columnsName : columnsNames) {
                    familiesQualifiersValues.add(Bytes.toBytes("f1"), Bytes.toBytes(columnsName), Bytes.toBytes(row.getAs(columnsName).toString()));
                }
                return new Pair(new ByteArrayWrapper(Bytes.toBytes(UUID.randomUUID().toString())), familiesQualifiersValues);
            }
        });

        hbaseContext.bulkLoadThinRows(myrdd,
                TableName.valueOf(tableName),
                new MyBulkLoadFunction(),
                args[0],
                new HashMap(),
                false,
                HConstants.DEFAULT_MAX_FILE_SIZE);

        jsc.stop();
        spark.stop();

    }


    private static class MyBulkLoadFunction implements Function<Pair<ByteArrayWrapper, FamiliesQualifiersValues>, Pair<ByteArrayWrapper, FamiliesQualifiersValues>> {
        @Override
        public Pair<ByteArrayWrapper, FamiliesQualifiersValues> call(Pair<ByteArrayWrapper, FamiliesQualifiersValues> row) throws Exception {
            return row;
        }
    }


}



